File: python/problems/leetcode/0102_binary_tree_level_order_traversal/solution.py
"""LeetCode 102: Binary Tree Level Order Traversal."""

from __future__ import annotations

from collections import deque
from typing import Optional


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: Optional["TreeNode"] = None,
        right: Optional["TreeNode"] = None,
    ) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> list[list[int]]:
        res: list[list[int]] = []
        if root is None:
            return res
        q: deque[TreeNode] = deque([root])
        while q:
            level: list[int] = []
            for _ in range(len(q)):
                node = q.popleft()
                level.append(node.val)
                if node.left is not None:
                    q.append(node.left)
                if node.right is not None:
                    q.append(node.right)
            res.append(level)
        return res


if __name__ == "__main__":
    root = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
    assert Solution().levelOrder(root) == [[3], [9, 20], [15, 7]]
    assert Solution().levelOrder(None) == []
    print("0102 OK")

File: cpp/problems/leetcode/0102_binary_tree_level_order_traversal/solution.cpp
// LeetCode 102
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
    TreeNode(int v, TreeNode* l, TreeNode* r) : val(v), left(l), right(r) {}
};

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if (!root) return res;
        queue<TreeNode*> q;
        q.push(root);
        while (!q.empty()) {
            int sz = (int)q.size();
            vector<int> level;
            for (int i = 0; i < sz; ++i) {
                TreeNode* n = q.front();
                q.pop();
                level.push_back(n->val);
                if (n->left) q.push(n->left);
                if (n->right) q.push(n->right);
            }
            res.push_back(level);
        }
        return res;
    }
};

int main() {
    TreeNode n15(15), n7(7), n9(9), n20(20, &n15, &n7);
    TreeNode root(3, &n9, &n20);
    assert((Solution().levelOrder(&root) ==
            vector<vector<int>>{{3}, {9, 20}, {15, 7}}));
    assert(Solution().levelOrder(nullptr).empty());
    cout << "0102 OK" << endl;
    return 0;
}
